internal/mtls/cert_utils.go

```diff
@@ -0,0 +1,174 @@
+	pem.Encode(privKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(c.privKey),
+	})
+	c.PrivKeyPEM = privKeyPEM
+}
+
+func (c *CertificateGroup) parseSignedCertificate() error {
+	var err error
+	c.signedCert, err = x509.ParseCertificate(c.certBytes)
+	return err
+}
+
+// GetCertificate returns the certificate Group in tls.Certficicate format.
+func (c *CertificateGroup) GetCertificate() (tls.Certificate, error) {
+	return tls.X509KeyPair(c.certPEM.Bytes(), c.PrivKeyPEM.Bytes())
+}
+
+func (c *CertificateGroup) GetCert() *x509.Certificate {
+	return c.cert
+}
+
+func getCACertificate() (*CertificateGroup, error) {
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(time.Now().Unix()),
+		Subject: pkix.Name{
+			Organization: []string{"K4e-operator"},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	// create our private and public key
+	caPrivKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot generate CA Key")
+	}
+
+	// create the CA
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivKey.PublicKey, caPrivKey)
+	if err != nil {
+		return nil, err
+	}
+
+	certificateBundle := CertificateGroup{
+		cert:      ca,
+		privKey:   caPrivKey,
+		certBytes: caBytes,
+	}
+	certificateBundle.CreatePem()
+	certificateBundle.parseSignedCertificate()
+	return &certificateBundle, nil
+}
+
+func getKeyAndCSR(cert *x509.Certificate, caCert *CertificateGroup) (*CertificateGroup, error) {
+
+	certKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot generate cert Key")
+	}
+
+	// sign the cert by the CA
+	certBytes, err := x509.CreateCertificate(
+		rand.Reader, cert, caCert.cert, &certKey.PublicKey, caCert.privKey)
+	if err != nil {
+		return nil, err
+	}
+
+	certificateBundle := CertificateGroup{
+		cert:      cert,
+		privKey:   certKey,
+		certBytes: certBytes,
+	}
+	certificateBundle.CreatePem()
+	certificateBundle.parseSignedCertificate()
+	return &certificateBundle, nil
+}
+
+func getServerCertificate(dnsNames []string, localhostEnabled bool, CACert *CertificateGroup) (*CertificateGroup, error) {
+
+	ips := []net.IP{}
+	if localhostEnabled {
+		ips = append(ips, net.ParseIP("127.0.0.1"), net.ParseIP("::1"))
+	}
+
+	cert := &x509.Certificate{
+		SerialNumber: CACert.cert.SerialNumber,
+		Subject: pkix.Name{
+			CommonName:   "*", // CommonName match all, and using ASN names
+			Organization: []string{serverCertOrganization},
+			Country:      []string{"US"},
+		},
+		DNSNames:     dnsNames,
+		IPAddresses:  ips,
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().AddDate(certDefaultExpiration, 0, 0),
```

Comment:
Do I understand correctly that every time we start the operator, the server certificate expiration changes? Meaning - it would have the same serial number and other details but different expiration every time?

Reply:
at some point, we're going to change this logic to retrieve the cert from a 3party provider, but It's the easiest way that I found right now.

Reply:
Do we create new cert and key upon restart even if it exists?

Reply:
mark the PR as WIP and I'll implement the configmap retrieval for this

Reply:
It seems we call InitCertificates during startup. Eventually, getServerCertificate is called. It creates a new CA cert. Is this correct? Don't we store the CA somewhere?

Reply:
CA cert is stored, TLS server certificate no.

Reply:
TLS server certificate is the one used by the operator? I mean the HTTP server within the operator? and we create a new one after every startup?

Reply:
Jira issue to track this: 
https://issues.redhat.com/browse/ECOPROJECT-431